gallery_dl/extractor/fantia.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+"""Extractors for https://fantia.jp/"""
+
+from .common import Extractor, Message
+from .. import text
+
+
+class FantiaExtractor(Extractor):
+    """Base class for fantia extractors"""
+    category = "fantia"
+    root = "https://fantia.jp"
+    directory_fmt = ("{category}", "{fanclub_id}")
+    filename_fmt = "{post_id}_{file_id}.{extension}"
+    archive_fmt = "{post_id}_{file_id}"
+    _warning = True
+
+    def items(self):
+        yield Message.Version, 1
+
+        if self._warning:
+            if "_session_id" not in self.session.cookies:
+                self.log.warning("no '_session_id' cookie set")
+            FantiaExtractor._warning = False
+
+        for url, data in self.posts():
+            if data["content_filename"]:
+                fname = data["content_filename"]
+                data["filename"] = fname
+                if "." in fname:
+                    data["extension"] = fname[fname.rfind(".")+1:]
+                else:
+                    data["extension"] = text.ext_from_url(url)
+            else:
+                data = text.nameext_from_url(url, data)
+            data["file_url"] = url
+
+            yield Message.Directory, data
+            yield Message.Url, url, data
+
+    def posts(self):
+        """Return all relevant post objects"""
+
+    def _pagination(self, base_url):
+        headers = {"Referer": self.root}
+        page = 1
+        posts_found = True
+
+        while posts_found:
+            url = base_url+str(page)
+            url = text.ensure_http_scheme(url)
+            gallery_page_html = self.request(url, headers=headers).text
+            posts_found = False
+            for post_id in text.extract_iter(
+                gallery_page_html, 'class="link-block" href="/posts/', '"'
+            ):
+                posts_found = True
+                for url, data in self._get_post_data_and_urls(post_id):
+                    yield url, data
```

Comment:
`yield from` like before, but it'd be better to split `_get_post_data_and_urls()`